File: src/lib/events.ts
/**
 * Centralised Tauri event names.
 *
 * Each entry is the literal string the backend uses with
 * `app_handle.emit(…)` or `Window::emit(…)`. Importing from here
 * (rather than typing the string at every call site) means a typo
 * is a TypeScript error instead of a silent listener that never
 * fires, and a rename touches one line instead of grep-and-pray.
 *
 * The constants are an immutable object literal rather than an
 * enum so call sites can pass them straight to `listen` / `emit`
 * without `as string` ceremony.
 *
 * Backend-side: these names are also baked into Rust strings.
 * Keep the two sides in sync; the four-place IPC sync rule from
 * `CLAUDE.md` applies the same way to events.
 */
export const Events = {
  /// Backend → frontend: hotkey toggle (`Ctrl+⌥/Alt+H` by default)
  /// fired. Frontend dispatches start-vs-stop based on its own
  /// recording state. See `+page.svelte`.
  HotkeyToggle: "hotkey:toggle",
  /// Backend → frontend: push-to-talk key down. Start dictation.
  HotkeyPttPress: "hotkey:ptt-press",
  /// Backend → frontend: push-to-talk key up. Stop dictation.
  HotkeyPttRelease: "hotkey:ptt-release",
  /// Frontend → frontend (broadcast): the main window's recording
  /// state changed. The HUD window listens to drive its visibility.
  UiRecordingState: "ui:recording-state",
  /// Backend → HUD window: the audio capture's RMS level for the
  /// last frame, in 0..1. Drives the HUD pill's level meter.
  AudioLevel: "audio:level",
  /// Backend → frontend (main): user picked a section from the
  /// native macOS menu bar. Payload is `"dictation" | "meetings" |
  /// "history"`. Sidebar uses this to flip the active tab.
  MenuGotoSection: "menu:goto-section",
  /// Frontend → settings window (broadcast): which tab to surface
  /// when Settings opens (or, if it's already open, switch to).
  SettingsGotoTab: "settings:goto-tab",
  /// Backend → all windows (broadcast): a model download is in
  /// flight. Payload is `{ id, received, total | null }`. The
  /// settings window's picker drives the progress bar.
  ModelDownloadProgress: "model:download-progress",
  /// Backend → all windows (broadcast): a model download finished
  /// successfully. Payload is `{ id }`. Main window uses this to
  /// clear its "no model installed" banner.
  ModelDownloadDone: "model:download-done",
  /// Backend → all windows (broadcast): a model download failed.
  /// Payload is `{ id, error }`. Settings window surfaces the
  /// error chip on the matching card.
  ModelDownloadFailed: "model:download-failed",
  /// Backend → frontend (main): a meeting-mode source (mic /
  /// system-audio) stopped capturing mid-session — TCC revoke,
  /// device unplug, etc. Payload is `{ kind: "microphone" |
  /// "system-audio", reason }`. Surfaces a struck-through chip
  /// in the active-session source line.
  MeetingSourceFailed: "meeting:source-failed",
  /// Backend → frontend (main): a mic source was lost mid-session
  /// and the pump has either fallen back to the system default or
  /// has no fallback. Payload: `{ sessionId, sourceKind, lostDevice,
  /// newDevice? }`. The meeting panel shows a banner with the lost
  /// device name and (if present) the fallback device name.
  AudioDeviceLost: "audio:device-lost",
  /// Backend → frontend (main): the original mic was detected on
  /// replug and the pump has swapped back. Payload: `{ sessionId,
  /// sourceKind, restoredDevice }`. Dismisses the device-lost banner.
  AudioDeviceRestored: "audio:device-restored",
  /// Backend → settings window: result of a Check for Updates
  /// probe fired from the macOS menu (#265). Payload is the
  /// `UpdateCheckResult` tagged union. The Settings About tab
  /// listens for this to render the result inline when the
  /// probe was triggered from the menu rather than the in-tab
  /// button.
  UpdaterResult: "updater:result",
  /// Backend → HUD window: lifecycle state for the recording
  /// HUD overlay (#291). Payload is `"recording"` or
  /// `"processing"`. Drives the HUD's render branch so the user
  /// sees a continuous "still working" signal across the
  /// transcription gap, instead of the HUD vanishing before
  /// the clipboard is updated.
  HudState: "hud:state",
  /// Frontend → all windows (broadcast): user picked an explicit
  /// theme (or reverted to System) in Settings → General →
  /// Appearance (#411 phase A). Payload is `"system" | "light"
  /// | "dark"`. Every window's root layout listens and re-applies
  /// the `data-theme` attribute on `<html>`. Canonical helpers
  /// live in `lib/theme.ts`.
  Theme: "hush:theme",
  /// Backend (profile auto-activate poller) → main window: a per-app
  /// audio profile (#427 Item 5) just activated because focus moved to
  /// an app with a populated profile. Fires only on transitions
  /// (the just-focused-app's profile differs from the last
  /// activation). Payload: `{ appName, preferredAudioSource,
  /// preferredModelId }`. Main window listener swaps the active
  /// source selector + invokes `model_select` for the model
  /// swap, then surfaces a transient "Switched to <app> profile."
  /// notice. Auto-apply is gated on `recording === false` so a
  /// mid-dictation focus change doesn't interrupt the active
  /// stream.
  AppProfileActivated: "app:profile-activated",
  /// Frontend → all windows (broadcast): user toggled the F5
  /// technical status line under the waveform (Settings → General
  /// → Advanced). Payload is `boolean`. The main window's
  /// ControlsSection listens and re-renders. Persistence is
  /// localStorage; this event only propagates *changes* across
  /// already-open windows. Canonical helpers live in
  /// `lib/status-line.ts`.
  StatusLine: "hush:status-line",
  /// Backend → settings window (debug tab): a new backend log
  /// entry is available. Payload is a `LogEntry` object. The
  /// DebugConsole component appends it to the in-page list.
  /// Subscribe *before* calling `get_log_entries` to guarantee no
  /// events are lost across the snapshot / live-stream gap (#532).
  LogEvent: "log:event",
  /// Backend → HUD window and main window: whisper.cpp inference
  /// progress during dictation transcription (integer 0–100).
  /// Throttled to every 5 percentage points in Rust to keep
  /// event-bus traffic low. The HUD shows "Processing… N%" in
  /// its label; the main window's RecordPanel renders a thin
  /// progress bar under the waveform during the transcribing
  /// phase (#566).
  TranscriptionProgress: "transcription:progress",
  /// Backend → all windows (broadcast): Hush's System Audio (Screen
  /// Recording TCC) permission was just confirmed via a real SCK
  /// probe after the user granted it in System Settings (#579).
  /// The main window listens to show the relaunch banner; other
  /// windows can use it to refresh permission state.
  ///
  /// Why a relaunch is needed: macOS caches the TCC deny in
  /// `mediaserverd`/`coreaudiod` for the lifetime of the current
  /// process — the grant takes effect only in a fresh process. See
  /// `learnings.md` for the full explanation.
  PermissionScreenRecordingGranted: "permission:screen-recording-granted",
} as const;

File: src-tauri/src/meeting/events.rs
//! Meeting-module event payloads and emission helpers.
//!
//! Centralizes the wire shapes passed through [`crate::events::EventEmitter`]
//! so lifecycle and pump code share one source of truth for event names and
//! payload fields.

use serde::Serialize;

use crate::events::EventEmitter;

fn emit_payload<T: Serialize>(event_emitter: &dyn EventEmitter, event: &str, payload: &T) {
    match serde_json::to_value(payload) {
        Ok(value) => event_emitter.emit_json(event, value),
        Err(error) => {
            tracing::warn!(
                error = ?error,
                event = %event,
                "EventEmitter: payload serialization failed; event dropped"
            );
        }
    }
}

/// Fired when the meeting pump drops a per-source capture path mid-session
/// (TCC revoke, device unplug, inference panic). Without this signal the
/// panel keeps showing "recording from mic + system audio" while one of
/// those sources has silently gone dead.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct MeetingSourceFailedPayload<'a> {
    pub session_id: i64,
    pub source_kind: &'a str,
    pub reason: &'a str,
    /// `true` when the failure came from `audio::DeviceLost` — the user's
    /// mic / AirPods disconnected mid-session or vanished during pre-warm.
    /// Lets the frontend branch on a typed flag instead of substring-
    /// matching `reason`, which #617 flagged as fragile to backend wording
    /// changes.
    pub device_lost: bool,
}

/// Tauri event name the pump fires when [`MeetingSourceFailedPayload`] is
/// the wire body. Centralized so the frontend's listener
/// (`Events.MeetingSourceFailed`) and the backend emit sites can't drift.
pub(super) const MEETING_SOURCE_FAILED_EVENT: &str = "meeting:source-failed";

/// Payload emitted by [`crate::meeting::SessionManager::start_manual`] when a
/// new session opens successfully (both manual button-press and HAL
/// auto-start paths). Centralized so the frontend's listener
/// (`Events.MeetingSessionStarted`) and every backend emit site stay in sync.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct MeetingSessionStartedPayload {
    pub session_id: i64,
}

/// Tauri event name for [`MeetingSessionStartedPayload`]. Matches the
/// TypeScript constant `Events.MeetingSessionStarted` in `events.ts`.
pub(super) const MEETING_SESSION_STARTED_EVENT: &str = "meeting:session-started";

/// Fired when the pump fails to persist a finished utterance to the DB.
/// Re-uses the `dictation:meeting-append-failed` wire name so the existing
/// frontend banner listener in `meeting-sessions.svelte.ts` picks it up
/// without a new event registration.  Payload: `{ error: String }`.
pub(super) const MEETING_APPEND_FAILED_EVENT: &str = "dictation:meeting-append-failed";

/// Payload for [`MEETING_APPEND_FAILED_EVENT`].
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct MeetingAppendFailedPayload {
    pub error: String,
}

pub(super) fn emit_utterance_append_failed(event_emitter: &dyn EventEmitter, error: &str) {
    emit_payload(
        event_emitter,
        MEETING_APPEND_FAILED_EVENT,
        &MeetingAppendFailedPayload {
            error: error.to_owned(),
        },
    );
}

/// Fired when a mic source is lost mid-session and the pump has switched to
/// the system default or has no fallback. Payload: [`AudioDeviceLostPayload`].
pub(super) const AUDIO_DEVICE_LOST_EVENT: &str = "audio:device-lost";

/// Fired when the original mic is detected on replug and the pump has swapped
/// back. Payload: [`AudioDeviceRestoredPayload`].
pub(super) const AUDIO_DEVICE_RESTORED_EVENT: &str = "audio:device-restored";

/// Payload for [`AUDIO_DEVICE_LOST_EVENT`].
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct AudioDeviceLostPayload<'a> {
    pub session_id: i64,
    pub source_kind: &'a str,
    pub lost_device: &'a str,
    /// `Some` when fallback succeeded (name of the device now recording).
    pub new_device: Option<&'a str>,
}

/// Payload for [`AUDIO_DEVICE_RESTORED_EVENT`].
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct AudioDeviceRestoredPayload<'a> {
    pub session_id: i64,
    pub source_kind: &'a str,
    pub restored_device: &'a str,
}

pub(super) fn emit_meeting_source_failed(
    event_emitter: &dyn EventEmitter,
    session_id: i64,
    source_kind: &str,
    reason: &str,
    device_lost: bool,
) {
    emit_payload(
        event_emitter,
        MEETING_SOURCE_FAILED_EVENT,
        &MeetingSourceFailedPayload {
            session_id,
            source_kind,
            reason,
            device_lost,
        },
    );
}

/// Fired when the meeting pump finishes (normal stop, auto-stop, or error).
/// Lets the frontend clear `meeting.activeId` even when the stop was backend-
/// driven (device failure, auto-stop) and no explicit `stopSession()` call
/// was made from the UI (#799).
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct MeetingSessionEndedPayload {
    pub session_id: i64,
}

/// Tauri event name for [`MeetingSessionEndedPayload`]. Matches the
/// TypeScript constant `Events.MeetingSessionEnded` in `events.ts`.
pub(super) const MEETING_SESSION_ENDED_EVENT: &str = "meeting:session-ended";

pub(super) fn emit_meeting_session_ended(event_emitter: &dyn EventEmitter, session_id: i64) {
    emit_payload(
        event_emitter,
        MEETING_SESSION_ENDED_EVENT,
        &MeetingSessionEndedPayload { session_id },
    );
}

pub(super) fn emit_meeting_session_started(event_emitter: &dyn EventEmitter, session_id: i64) {
    emit_payload(
        event_emitter,
        MEETING_SESSION_STARTED_EVENT,
        &MeetingSessionStartedPayload { session_id },
    );
}

pub(super) fn emit_audio_device_lost(
    event_emitter: &dyn EventEmitter,
    session_id: i64,
    source_kind: &str,
    lost_device: &str,
    new_device: Option<&str>,
) {
    emit_payload(
        event_emitter,
        AUDIO_DEVICE_LOST_EVENT,
        &AudioDeviceLostPayload {
            session_id,
            source_kind,
            lost_device,
            new_device,
        },
    );
}

pub(super) fn emit_audio_device_restored(
    event_emitter: &dyn EventEmitter,
    session_id: i64,
    source_kind: &str,
    restored_device: &str,
) {
    emit_payload(
        event_emitter,
        AUDIO_DEVICE_RESTORED_EVENT,
        &AudioDeviceRestoredPayload {
            session_id,
            source_kind,
            restored_device,
        },
    );
}
